src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java

```diff
@@ -208,6 +215,40 @@ public boolean accept(Path path) {
     }
 
     /**
+     * Ensure reads in a pair fall in the same partition (input split), if the reads are queryname-sorted,
+     * so they are processed together. No shuffle is needed.
+     */
+    JavaRDD<GATKRead> putPairsInSamePartition(final SAMFileHeader header, final JavaRDD<GATKRead> reads) {
```

Comment:
The trouble is that it is very difficult to include in Hadoop-BAM, since with the new optimized interval filtering in https://github.com/HadoopGenomics/Hadoop-BAM/pull/122 it's not possible to continue getting reads from the BAMFileReader when it's past the end of a chunk. The alternative - putting this RDD code into Hadoop-BAM - is a bigger change as Hadoop-BAM is an MR-level library so it doesn't have any Spark code in it. 

I would prefer to keep the code here so we can iterate on it (David has some suggested improvements). Perhaps we can add it to Hadoop-BAM in time.